### ISSUE TITLE
Case sensitive for matching email field, case insensitive for updating

### ISSUE THREAD

**Issue by iancrew** [adobe-apiplatform/user-sync.py#471]
For the email field of a user's info (note: NOT the username field), Adobe User Sync 2.4 does a case-sensitive match to determine if that field needs to be updated.

So, a user foobar@example.com that has email address "foobar@example.com" in the company's directory and "FooBar@EXAMPLE.COM" in the email field of their Federated ID in Adobe CC would be properly flagged as needing to have the email field of their Adobe account updated, even though those two addresses only differ by case.

When Adobe User Sync runs, it properly sends an update that looks like:

`2019-03-22.log:2019-03-22 05:17:02 18899 INFO processor - Updating info for user key: federatedID,foobar@example.com, changes: {'email': 'foobar@example.com'}`

That update appears to succeed (no error is returned). However, the email address isn't actually updated in the Adobe Admin Console, and the Adobe User Sync tool attempts to update the same email address again the next time it runs.

I believe what's happening is that the update code on the Adobe side of the APIs is doing a case-insensitive match at some point during the update procedure, and determining that the user's email address field doesn't need to be updated. For example, just as a guess, many SQL instances are case-insensitive, unless specifically set otherwise in either the table or the query.

Ideally, the updates would also be case-sensitive, so that the email field in Adobe would fully match the email field in our directory.

This is affecting 432 of the ~57K people in my Adobe CC Admin Console. (And no, I don't know how they got out of sync in that manner, sorry.)

**Comment by n8felton** [comment]
Just adding that I'm seeing this exact issue as well, for ~170 users out of ~27k in the console.

**Comment by Luci2015** [comment]
This is not an UST bug, rather a limitation of the dependent service that the UM API is calling to request the change.
Inside the UM API [documentation](https://adobe-apiplatform.github.io/umapi-documentation/en/api/ActionsCmds.html) it is stated that the update call cannot happen for the email on letter case changes:
> For all account types, the email address is used as a case-sensitive unique identifier. You cannot use the update request to change the case of the email address, once it has been set.

If it's really important for the email to be exact same case as the one in the LDAP, the only way to make this happen is either to change the letter case in the LDAP to match Admin Console or to delete the user from Directory Users menu and let the tool re-create the account with the mail value from the LDAP. Deleting the account from Directory Users menu will also involve **losing** the associated cloud stored data for that account (if any)!!

**Comment by adorton-adobe** [comment]
@Luci2015 You are correct in how the UMAPI handles letter casing for email addresses, but it looks to me like the UMAPI documentation might need to be reworded for clarity.  If email were truly case-sensitive, then **user@example.com** and **USER@EXAMPLE.COM** would be treated as totally different identifiers and could be used to identify two different users.  But that isn't the case.  It would be more accurate to state that the email address is case-insensitive.

I'm not even sure that it's accurate to state that it will retain the letter case given to it upon user creation.  I've seen plenty on instances where the email address will be converted to upper case when a new user is created, even though the email address was provided in lower or mixed case.  

Regardless, the sync tool should take this into account when comparing email addresses for user-info updates.  That is why I consider this a bug - the sync tool is designed to normalize letter casing when comparing email addresses.  This is an edge case that has been missed up to this point.

**Comment by Luci2015** [comment]
Normalisation is a good idea in the measure where you are not interested if the mail value in Admin Console matches exactly the one you have in AD for your user.
I don't know if there is any SSO impact if the mail value in Admin Console is different than the one in AD - that would make the Normalisation a problem because with current configuration you get at least the notification that an attempt to update the mail is made (but see it as a success even if nothing changes).
If there are any SSO ramifications on case sensitive mails, then maybe UST should trigger Warnings about found case mismatch on mails just like it does for missing attributes with an option to hide these log entries if you're not interested to see them? Just thinking out loud here..

**Comment by ancientHacker** [comment]
There is no bug here, except in the documentation.  Adobe-side handling of both usernames and email addresses are completely case-independent; however, user sync must *never* alter the case of email addresses sent when a user is created because that case is remembered on the Adobe side and the *customer's email system* may be sensitive to case.

I would recommend that the documentation be altered to say the following:

> The user sync tool never alters the case of a user's email as found in the customer directory, so that when a customer first creates a user in a directory via sync that user's email will match the case in the customer directory at that time.  This precaution is required because customer email systems are allowed to be sensitive to case in the local part of the email address.

> However, it's important to note that handling of all usernames and emails on the Adobe side is always case-_insensitive_, so that it is _not_ possible to have two users whose email differs only in case: on the Adobe side those would be seen as two users with the same email, which is not allowed.  Accordingly, when comparing ldap users with Adobe users, the user sync tool uses case-_insensitive_ comparison, so that its sync logic is not misled by alterations that have taken place on the customer directory side.

>As a result of the above, then, it is *not* possible to alter the case of an existing user's email on the Adobe side by specifying an update with the new case: the case of usernames and emails are never updated on the Adobe side.  If it is required to update the case of a user who already exists on the Adobe side, you must first *delete* the user from the Adobe directory (which will remove that user from all organizations and trigger asset deletion - or reclamation in the case of ESM), and then you must create a new user in the Adobe directory the case of whose email is as desired.

**Comment by iancrew** [comment]
There most definitely *is* a bug here. More than one, I’d argue:

1)  The User Sync tool does a case-sensitive match when determining whether an update is necessary. It therefore attempts to do an update for those addresses every time it runs. But the current version of the Adobe APIs are case-insensitive, so the record never gets updated, resulting in spurious calls every single time the User Sync tool runs. 

2) There’s a bug in the Adobe APIs by being case-insensitive. I, as the paying customer, *must* be assumed to be the source of truth about my users’ information. If I tell Adobe to update a user’s account from FooBar@EXAMPLE.COM to foobar@example.com, why should Adobe be telling me that that’s not allowed/not correct?

We, as the central IT team running User Sync know what our organization’s own internal records say infinitely better than Adobe does, and my organization owns and controls all of the addresses in our domain(s). Also, the users are my organization’s users, not Adobe’s. My organization is Adobe’s customer, not the individual users. 

And no, deleting and recreating the account isn’t an option, as that user may have data stored in their Adobe CC account.

**Comment by Luci2015** [comment]
Did a test in Admin Console today and it seems that even the delete of the account from Directory Users and re-creation either with the API or manually in Admin Console fails in creating the account with the new letter case. Will dig this some more, as I remember it succeeding in the past.

**Comment by iancrew** [comment]
@Luci2015 yeah, I also tried doing a rename as follows, which also didn't work:

1) FooBar@EXAMPLE.COM -> FooBar-rename@EXAMPLE.COM
2) FooBar-rename@EXAMPLE.COM -> foobar@example.com

The first of those renames did appear to work, but when I did the second, the name reverted to FooBar@EXAMPLE.COM. So something definitely weird is going on here.

**Comment by adobeDan** [comment]
@iancrew @luci2015 This looks like a new bug has appeared in UMAPI: both these renames should work. To make sure I give that team enough info can you please: tell me whether these are federated orgs, give me the org IDs you are testing against, and any request IDs you are seeing in your logs?  I will also try to repro. @adorton you should too.

**Comment by adobeDan** [comment]
@iancrew While I may not agree with your analysis about whether either user sync or UMAPI has a bug, I agree that there should be a non-destructive way to change the case of the email of an existing user (first by renaming to a temp email, then changing It back with the correct case.

**Comment by iancrew** [comment]
@AdobeDan:

1) I doubt I have any logs from the experiments I did with the renaming—I did them with CSV files back around the time I submitted the bug. These tests were with Federated IDs, yes. I’m not sure where I’d find the OrgID—I’m happy to provide it if you can point me in the right direction. 

2) I’m sorry to disagree so vehemently, but this most definitely is a bug in UMAPI, and possibly also in User Sync. There’s absolutely zero reason to force the team managing AdobeCC for an organization to jump through all sorts of weird workarounds just to make the records in Adobe’s systems match the definitive source of truth (the customer’s own records). The “rename to something different, then change it back” is an OK kludge to workaround the bug for now, if it worked, but it’s not a fix. This is one of those instances where “the customer is always right” is very true.

**Comment by adorton-adobe** [comment]
I've updated the user update attribute check to normalize the email address to prevent unnecessary API calls.

**Comment by adorton-adobe** [comment]
@iancrew This fix only applies to the sync tool.  I can try to bring this to the attention of the platform team if I can reproduce it.  But for now, the sync tool will ignore case for email comparison when deciding which attributes need updating.

**Comment by Rob-StClairCollege** [comment]
We've also experienced the same mentioned above. Version 2.5.0rc3 seemed to have resolved it (thank you) but we've noticed a new issue with this release. If the username attribute (in our case it's mapped to the user's UPN) is different than the mail attribute, the user account doesn't get deleted. UST indicates in the logs that it has deleted but the change isn't reflected on the Adobe side. However, if UPN and mail attribute do indeed match, then no issues. Anyone else encountering the same problem?
